### DISCUSSION TITLE
[Roadmap] Astro v1.0.0 Beta => Stable

### DISCUSSION THREAD

**Discussion by natemoo-re** [withastro/roadmap#161] [Proposal]
# Summary

> **Status: Out of Date**  
> This list was used in the early days of the v1.0.0 beta, but it's no longer up to date! The current Astro v1 Beta is close to feature complete now, with no large tasks remaining and no large tasks on deck until v1.0.0 has been released. Visit https://docs.astro.build/ to learn what features are currently supported in Astro.

Astro v1.0.0 is now in Beta! We've already done a ton of work to [finalizing our APIs](https://github.com/withastro/rfcs/discussions/1) and implement breaking changes, but there's still more to do before a stable v1.0.0!

This document serves as a place to outline remaining v1.0.0 tasks, so that we can share some insight into what we'll be working on during this v1.0.0 Beta period.

## ✅ Address private/internal APIs

<details>
  <summary>Expand</summary>

- [x] Owner: @natemoo-re
- [x] Finalize `Astro.slots` API
- [x] Make `__renderMarkdown` non-enumerable
- [x] Move from `data-astro-raw` to `is:raw` directive

</details>

## ✅ AST output from `@astrojs/compiler`

<details>
  <summary>Expand</summary>

- [x] Owner: @natemoo-re 
- [x] Implementation: [PR](https://github.com/withastro/compiler/pull/274)

</details>

## ⏹ Render Hooks in Integrations

This is pretty tough, since it will force us to figure out how to add an integration that runs in SSR (currently, all integrations run during your build only).

- [ ] Owner: TBD
- [ ] Implementation: TBD

## ⏹ SSR API Improvements & Sugar

Currently, the SSR API is pretty low-level. For example, do we want [helpers for caching](https://nextjs.org/docs/api-reference/data-fetching/get-static-props#revalidate)? 

- [ ] Owner: TBD
- [ ] Implementation: TBD

## ⏹ Turbo Mode

SPA mode that Nate has been playing around with, to replace the deprecated `turbolinks`

- [ ] Owner: TBD
- [ ] Implementation: TBD

## ⏹ Benchmarks

We'd like to automate benchmarks of Astro's build performance and continue to improve our speed.

- [ ] Owner: TBD
- [ ] Implementation: TBD

## ⏹ Output Audit, Performance Improvements

Astro's output still has some low-hanging fruit to optimize performance, such as automatic prefetching of assets, removing render-blocking resources, and more.

- [ ] Owner: TBD
- [ ] Implementation: TBD

## ⏹ Astro Directives documentation

As a larger push to improve our documentation for v1.0.0, we'd like a single document that outlines all available Astro Directives

- [x] Owner: TBD
- [x] Implementation: TBD

## ⏹ RSS Feed Finalization

What do we do with `rss()`?

- [ ] Owner: TBD
- [ ] Implementation: TBD

## ⏹ Alias Audit

`~/` as an alias to the `src` directory.

- [ ] Owner: TBD
- [ ] Implementation: TBD

## ⏹ Markdown Finalization

- People don't like copy-pasting the same `layout` line into 100s of pages.  Is this possible? Can we deliver it without introducing too much implicit data flow?
- `<Markdown>` is forcing us to handle Markdown compilation at runtime. This breaks or degrades SSR. Should we remove it in favor of `import` + `.md` files?

- [ ] Owner: TBD
- [ ] Implementation: TBD

## Out of Scope (post-v1.0)

- **exports in components** (other than the currently-supported `getStaticPaths()` function)
- **incremental builds**
- **custom user directives**
- **image component** (vite-imagetools as v1.0 solution?)
- **Shared context setup: React, Svelte, Vue Plugins, etc.**
- **Astro.createComponent() API** (or similar)

**Comment by RafidMuhymin** [comment]
> `<Markdown>` is forcing us to handle Markdown compilation at runtime. This breaks or degrades SSR. Should we remove it in favor of import + .md files?

If we want to do this, we have to first improve the support for using components and dynamic values inside `.md` files. When writing docs for `astro-imagetools`, I faced a ton of issues and had to escape many times to use components and dynamic values.

**Comment by FredKSchott** [comment]
Notes from RFC Call:

#### `local:*` directive 
out of scope for v1.0, too much still to discuss

#### `asset/` magic folder
to be explored before v1.0, still issues that users are hitting without this

#### finalizing our default folders

- `static` vs. `public`
- `src/pages` vs. `src/routes`
- related: do we have an `assets/` folder?
- needs to be non-breaking if we do this

#### Turbo/SPA mode notes

- HTML fragments should be supported in `src/pages` today (we prepend DOCTYPE, problem?)
- could we only add DOCTYPE if we see a `html`, `head`, or `body` element?
- or, API endpoints that can return HTML

#### Fetch polish & finalization

- `fetch()` doesn't work for local files, intentional?
- `fetch()` different behavior in dev vs. build
- `fetch()` different behavior in ssg (fetch doesn't work) vs. ssr (fetch works)
- help users if they try to do the wrong thing (fetch `/...` from ssg)

#### Markdown

- Component alt: `<Markdown content={`...`} />`
- global config for markdown?
- default imports of components, without an import or `setup` script, use `<Code>` or `<TweetEmbed>`

#### Finalize Integrations, SSR, Adapters

- remove `--experimental` flags
- remove warnings for 3rd party integrations
- adding `astro add` support for 3rd party integrations
- can integrations prompt user for questions / run scripts after adding to the project
- finalize proper `package.json` settings for integrations (ex: `astro-component` for keywords?)

**Reply by FredKSchott** [reply]
#### Finalize Lit

- lots of complexity for Lit being the only framework that tries to own `.js` 
- can we force Lit components to use `.lit.js` instead?

**Comment by RafidMuhymin** [comment]
If the compiler handled the transformations of a sugar/non-existent `<Markdown />` component instead of an actual Astro component, would it solve the issues with SSR?

**Comment by inwardmovement** [comment]
> People don't like copy-pasting the same `layout` line into 100s of pages.  Is this possible? Can we deliver it without introducing too much implicit data flow?

Hugo is handling this with [frontmatter cascade](https://gohugo.io/content-management/front-matter#front-matter-cascade)

**Comment by janat08** [comment]
Prefetching assets means js/html too right?

**Reply by janat08** [reply]
In related issue this was mentioned: https://www.npmjs.com/package/astro-spa

**Reply by janat08** [reply]
Which has severe outstanding issues, but also decent feature set.

**Comment by inwardmovement** [comment]
Is this still used to track progress until end of july?

**Reply by FredKSchott** [reply]
Oh wow, I forgot about this one! Nope, this is no longer up to date. There's really nothing big left other than bug fixing and polish. We're still targeting end of July, although with some of the team out this month this may push out to the first or second week of August. We'll have one final update out message out once we have a final release date agreed on!

**Comment by matthewp** [comment]
Close?

**Reply by jasikpark** [reply]
*Lock :)
